### PR TITLE
Kemp/curvature validation

### PR DESCRIPTION
**Pull request by tdkemp**

### CODE REVIEW

**Comment by tdkemp** [comment]
@Besler I'd love your review on these additions. Thanks!

File: tests/layer/test_dice.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        g1.data[1, 1] = 1
+
+        g2 = g1.Copy()
+
+        L = Dice()
+        result = L(g1, g2)
+
+        self.assertEqual(result.data.shape, ())
+
+        expected_result = 1.0
```
Comment (Besler) [inline tests/layer/test_dice.py:28]: Testing that Dice(A, A) = 1.0 is great. We could add two more tests here to be sure.

Dice(1, 0) = 0.
Dice(A, B) = known value where A is say a sphere and B a sphere slightly shrunk. Something we know. If you trust your results for publication, we can go with it.

File: tests/layer/test_surface_distance.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+        g2 = g1.Copy()
+
+        D = SurfaceDistance()
+        result = D(g1, g2)
+
+        self.assertEqual(result.data.shape, ())
+
+        expected_result = 0.0
+
```
Comment (Besler) [inline tests/layer/test_surface_distance.py:29]: See Dice

File: tests/layer/test_jaccard.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        g1.data[1, 1] = 1
+
+        g2 = g1.Copy()
+
+        L = Jaccard()
+        result = L(g1, g2)
+
+        self.assertEqual(result.data.shape, ())
+
+        expected_result = 1.0
```
Comment (Besler) [inline tests/layer/test_jaccard.py:28]: see dice

File: tests/layer/test_total_distance.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        g1.data[1, 1] = 1
+
+        g2 = g1.Copy()
+
+        D = TotalDistance()
+        result = D(g1, g2)
+
+        self.assertEqual(result.data.shape, ())
+
+        expected_result = 0.0
```
Comment (Besler) [inline tests/layer/test_total_distance.py:49]: See dice

File: wolff/layer/jaccard.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        intersection = VI.volume
+
+        union = Union()(grid1, grid2)
+        VU = Volume()
+        VU(union)
+        union = VU.volume
+
+        jaccard = intersection/union
+
+        return jaccard
```
Comment (Besler) [inline wolff/layer/jaccard.py:29]: Some really beautiful code here and w/ Dice

File: wolff/layer/surface_distance.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class SurfaceDistance(Layer):
+    '''Compute the Dice coefficient between two level sets'''
+
+    def __init__(self):
+        super(SurfaceDistance, self).__init__()
+
+    def process(self, grid1, grid2, **kwargs):
+        # Compute Intersection
```
Comment (Besler) [inline wolff/layer/surface_distance.py:14]: Remove comment?

File: wolff/layer/total_distance.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            decay_rate=decay_rate,
+            method='sin'
+        )(gridA)
+
+        # Compute magnitude gradient
+        mag_grad = np.linalg.norm(np.stack(
+            np.gradient(gridA.data, *gridA.spacing)
+        ), axis=0)
+
+        e = np.abs(gridB.data-gridA.data)
```
Comment (Besler) [inline wolff/layer/total_distance.py:30]: I would add a comment here saying "this makes the `TotalDistance` a valid metric. Alternatively, we need to resolve the curve between grid points" or something similar.

File: wolff/sources/sphere_distance_transform.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        # Compute an image of difference vectors from center
+        diff_vector = np.zeros_like(index)
+        for i in range(index.shape[0]):
+            diff_vector[i] = index[i] - self.center[i]
+
+        # Now, create the perpendicular vector and take the magnitude
+        result = np.zeros(index.shape[1:])
+        for i in range(index.shape[0]):
+            result += np.square(diff_vector[i])
+        return (np.sqrt(result) - self.radius)
```
Comment (Besler) [inline wolff/sources/sphere_distance_transform.py:30]: See sphere source.

File: wolff/sources/sphere.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        # Compute an image of difference vectors from center
+        diff_vector = np.zeros_like(index)
+        for i in range(index.shape[0]):
+            diff_vector[i] = index[i] - self.center[i]
+
+        # Now, create the perpendicular vector and take the magnitude
+        result = np.zeros(index.shape[1:])
+        for i in range(index.shape[0]):
+            result += np.square(diff_vector[i])
+        return result <= np.square(self.radius)
```
Comment (Besler) [inline wolff/sources/sphere.py:30]: You can do this more efficiently like this:
```python
        # Create sphere
        d = 2
        size = 100
        r = 25
        center = np.array([size/2.0]*d)
        I = np.indices([size]*d)

        phi = I - np.expand_dims(center, list(range(1, d+1)))
        phi = np.linalg.norm(phi, axis=0) - r
```
But I don't care.

**Review by Besler — APPROVED** [review]
Some really excellent stuff here. Well done.

**Pull request merged**